### PR TITLE
Add support for `Astro.currentLocale`

### PR DESCRIPTION
**Pull request by HiDeoo**
#### Description

- Closes #1176

This PR adds support for using [`Astro.currentLocale`](https://docs.astro.build/en/reference/api-reference/#astrocurrentlocale) in Starlight, and also technically adds support for using an [Astro i18n configuration](https://docs.astro.build/en/reference/configuration-reference/#i18n) instead of a [Starlight i18n configuration](https://starlight.astro.build/guides/i18n/).

First, a new "concept" of Starlight built-in default locale has been introduced. This matches the current English one, except now it's something generated once, based on a BCP 47 tag, and can be used everywhere in the codebase instead of hardcoding `'en'`, `'English'` or `'ltr'` everywhere.

Then, 4 main possibilities are supported when processing both i18n configurations:

- If no Astro and Starlight i18n configurations are provided, the built-in default locale is used in Starlight and a matching Astro i18n configuration is generated/used.
- If only a Starlight i18n configuration is provided, an equivalent Astro i18n configuration is generated/used.
- If only an Astro i18n configuration is provided, the Starlight i18n configuration is updated to match it.
- If both an Astro and Starlight i18n configurations are provided, an error is thrown.

A few additional details:

- The minimum required Astro version has been updated to ~~`v4.4.5`~~ `v4.8.0` just like #1846 so that the following PRs can also be relied upon:
  - https://github.com/withastro/astro/pull/10169
  - https://github.com/withastro/astro/pull/10196
  - https://github.com/withastro/astro/pull/10193
- Some tests may look a bit repetitive, could definitely be refactored to rely more on `test.each`, etc. but I think considering it's a lot of data massaging with internal shapes, with pretty different and identified cases, I personally found that the extra verbosity was a bit better for readability in this case.
- The links validation is failing a lot but this is expected as the links validator plugin does not support any of these changes.

Regarding the use of an Astro i18n configuration, some inference is done when it comes to locale data, such as labels and text direction. This is due to the fact that Astro doesn't provide a way to specify these in the configuration, so they are inferred using [`Intl`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) features compatible with the minimum Node.js version supported by Astro.

For comparison, in the Starlight documentation, here is a language selector labels comparison between the current Starlight i18n configuration and the equivalent inferred Astro i18n configuration:

| Starlight i18n config | Astro i18n config |
| --------------------- | ----------------- |
| SCR-20240508-jgwp       | SCR-20240508-jhmh   |

There are some differences but the code is pretty similar to #1012 where we already got a lot of approvals, so I would think this should be fine, and if needed, people can always switch to the Starlight i18n configuration for more control.

#### Remaining tasks

- [x] Switch back to the Starlight i18n configuration in the documentation
- [x] Remove the `Astro.currentLocale` debug infos in `packages/starlight/components/Banner.astro`
- [x] Figure out what and how much should be documented
- [x] Write the documentation
- [x] Add a changeset

### CODE REVIEW

File: packages/starlight/package.json [hunk 1]
```diff
@@ -180,15 +180,15 @@
...
   "devDependencies": {
     "@astrojs/markdown-remark": "^4.2.1",
     "@playwright/test": "^1.43.1",
     "@types/node": "^18.16.19",
-    "@vitest/coverage-v8": "^1.3.1",
-    "astro": "^4.3.5",
-    "vitest": "^1.3.1"
+    "@vitest/coverage-v8": "^1.4.0",
+    "astro": "^4.4.5",
+    "vitest": "^1.4.0"
```
Comment (HiDeoo) [inline packages/starlight/package.json:191]: The Vitest update is required due to https://github.com/withastro/astro/pull/10120 shipped in Astro `v4.4.0`.

File: packages/starlight/utils/translations.ts [hunk 2]
```diff
@@ -4,7 +4,9 @@ import type { i18nSchemaOutput } from '../schemas/i18n';
 import { createTranslationSystem } from './createTranslationSystem';
 
 type UserI18nSchema = 'i18n' extends DataCollectionKey
-	? CollectionEntry<'i18n'>['data']
+	? CollectionEntry<'i18n'> extends { data: infer T }
```
Comment (HiDeoo) [inline packages/starlight/utils/translations.ts:7]: This new change is required to due this [change](https://github.com/withastro/astro/pull/10001) in Astro `v4.4.0` which makes CC types include collection names even when a matching directory is absent.

**Comment by HiDeoo** [comment]
Updated the PR, bumped the required Astro version to 4.8.0 to match #1846 and also updated the code to handle the Astro i18n [manual routing strategy](https://github.com/withastro/astro/pull/10193) added in Astro 4.6.0.

File: packages/starlight/utils/i18n.ts [hunk 3]
```diff
@@ -1,3 +1,184 @@
...
+/** Check if the Starlight built-in default locale is used in a Starlight config. */
+function isUsingBuiltInDefaultLocale(config: StarlightConfig) {
+	return (
+		!config.isMultilingual &&
+		config.locales === undefined &&
+		config.defaultLocale.label === BuiltInDefaultLocale.label &&
+		config.defaultLocale.lang === BuiltInDefaultLocale.lang &&
+		config.defaultLocale.dir === BuiltInDefaultLocale.dir
+	);
+}
```
[the remark this answers is not in the document]
Comment (HiDeoo) [reply packages/starlight/utils/i18n.ts]: I love it. I was not super happy about this function but totally missed that opportunity, thanks a lot.

File: packages/starlight/utils/i18n.ts [hunk 4]
```diff
@@ -1,3 +1,184 @@
...
+			'Cannot provide both an Astro i18n configuration and a Starlight i18n configuration.',
+			'Remove one of the i18n configurations.\nSee more at https://starlight.astro.build/guides/i18n/'
+		);
+	} else if (astroI18nConfig) {
+		if (astroI18nConfig.fallback) {
+			throw new AstroError(
+				'Starlight is not compatible with the `fallback` option in the Astro i18n configuration.',
+				'Starlight uses its own fallback strategy showing readers content for a missing page in the default language.\nSee more at https://starlight.astro.build/guides/i18n/#fallback-content'
+			);
+		}
```
[the remark this answers is not in the document]
Comment (HiDeoo) [reply packages/starlight/utils/i18n.ts]: Well, I guess this hints that my wording is not good at all and the decision to have this check too maybe.

You are totally right and my idea was that a user could maybe think that the Astro `fallback` option could control Starlight behavior too. But I guess indeed, the use case you mentioned could be a valid one too so I guess we should just remove this check.

File: packages/starlight/utils/i18n.ts [hunk 5]
```diff
@@ -1,3 +1,184 @@
...
+
+/** Returns the locale informations such as a label and a direction based on a BCP-47 tag. */
+function getLocaleInfo(lang: string) {
+	try {
+		const locale = new Intl.Locale(lang);
+		const label = new Intl.DisplayNames(locale, { type: 'language' }).of(lang);
+		if (!label || lang === label) throw new Error('Label not found.');
+		return {
+			label: label[0]?.toLocaleUpperCase(locale) + label.slice(1),
+			// `textInfo` is not part of the `Intl.Locale` type but is available in Node.js 18.0.0+.
```
[the remark this answers is not in the document]
Comment (HiDeoo) [reply packages/starlight/utils/i18n.ts]: Good idea, altho we still need a cast to explicitly get `'ltr' | 'rtl'`.

File: packages/starlight/utils/i18n.ts [hunk 6]
```diff
@@ -1,3 +1,184 @@
...
+					inferStarlightLocaleFromAstroLocale(locale),
+				])
+		  );
+
+	const defaultAstroLocale = astroI18nConfig.locales.find((locale) =>
+		isDefaultAstroLocale(astroI18nConfig, locale)
+	);
+
+	// This should never happen as Astro validation should prevent this case.
+	if (!defaultAstroLocale) throw new Error('Astro default locale not found.');
```
[the remark this answers is not in the document]
Comment (HiDeoo) [reply packages/starlight/utils/i18n.ts]: Neat, I only added the bug report template to the URL query params if that's fine.

**Comment by HiDeoo** [comment]
Thanks a lot for the great review as usual, super appreciated and very helpful! :raised_hands:

Updated the PR with the suggested changes and also took a first stab at documenting the new changes.

File: docs/src/content/docs/guides/i18n.mdx [hunk 7]
```diff
@@ -272,3 +274,17 @@ export const collections = {
 ```
 
 Learn more about content collection schemas in [“Defining a collection schema”](https://docs.astro.build/en/guides/content-collections/#defining-a-collection-schema) in the Astro docs.
+
+## Accessing the current locale
+
+You can read the current locale from the `Astro` global that is available in all contexts in `.astro` files with the [`currentLocale`](https://docs.astro.build/en/reference/api-reference/#astrocurrentlocale) property.
+
+The following example reads the current locale and uses it to generate a link to an About page in the current language:
```
Comment (delucis) [inline docs/src/content/docs/guides/i18n.mdx:282]:
```suggestion
You can use [`Astro.currentLocale`](https://docs.astro.build/en/reference/api-reference/#astrocurrentlocale) to read the current locale in `.astro` components.

The following example reads the current locale and uses it to generate a link to an about page in the current language:
```

**Review by delucis — COMMENTED** [review]
Left a small suggestion to simplify the docs, but otherwise I think this looks good to me! We can always revisit documentation later once we see what people are finding unclear.

I’ll take care of reverting the TODOs now.

**Review by delucis — APPROVED** [review]
Approving! These docs work for me, and we will have an opportunity to expand on them probably when working on #1923.

Thanks again @HiDeoo! Stellar work 🌟

**Pull request merged**